store/types/listening.go

```diff
@@ -0,0 +1,42 @@
+package types
+
+import (
+	"io"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+)
+
+// WriteListener interface for streaming data out from a listenkv.Store
+type WriteListener interface {
+	// if value is nil then it was deleted
+	// storeKey indicates the source KVStore, to facilitate using the the same WriteListener across separate KVStores
+	// set bool indicates if it was a set; true: set, false: delete
+	OnWrite(storeKey StoreKey, set bool, key []byte, value []byte)
+}
+
+// StoreKVPairWriteListener is used to configure listening to a KVStore by writing out length-prefixed
+// protobuf encoded StoreKVPairs to an underlying io.Writer
+type StoreKVPairWriteListener struct {
+	writer     io.Writer
+	marshaller codec.BinaryMarshaler
+}
+
+// NewStoreKVPairWriteListener wraps creates a StoreKVPairWriteListener with a provdied io.Writer and codec.BinaryMarshaler
+func NewStoreKVPairWriteListener(w io.Writer, m codec.BinaryMarshaler) *StoreKVPairWriteListener {
+	return &StoreKVPairWriteListener{
+		writer:     w,
+		marshaller: m,
+	}
+}
+
+// OnWrite satisfies the WriteListener interface by writing length-prefixed protobuf encoded StoreKVPairs
+func (wl *StoreKVPairWriteListener) OnWrite(storeKey StoreKey, set bool, key []byte, value []byte) {
+	kvPair := new(StoreKVPair)
+	kvPair.StoreKey = storeKey.Name()
+	kvPair.Set = set
+	kvPair.Key = key
+	kvPair.Value = value
+	if by, err := wl.marshaller.MarshalBinaryLengthPrefixed(kvPair); err == nil {
```

Comment:
Any particular reason we're length-prefixing here?

Reply:
We're length-encoding so that we can write all of these KVPairs out to the same stream or file without any other sort of delineation and be able to segment and decode the separate messages at a later time. I supposed the length prefixing could be handled at the streaming service level, which would allow the service to decide how they want to perform the delineation rather than requiring this length-prefixing approach. What do you think would be best?

Reply:
Got it, i think length-prefixing makes sense then.

Reply:
Protobuf supports streaming. Otherwise, this should be handled on the communication protocol level, not data serialization.
If such a strategy would be useful (doing length prefixing of every data), then it would be supported as an option by all major serialization mechanisms.

Can we revisit and not use `*LengthPrefixed`?